In [1]:
import numpy as np
import pandas as pd

# New Section

In [6]:
books=pd.read_csv('/content/Books.csv',encoding='latin-1')
ratings=pd.read_csv('/content/Ratings.csv',encoding='latin-1')
users=pd.read_csv('/content/Users.csv',encoding='latin-1')

<ipython-input-6-05c62ae7f807>:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books=pd.read_csv('/content/Books.csv',encoding='latin-1')


In [7]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [8]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [9]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [10]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [11]:
print(books.shape)
print(ratings.shape)
print(users.shape)

(271360, 8)
(1149780, 3)
(278858, 3)


In [12]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

It will give you the null value in the file

In [13]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [14]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

To check the duplicate values in books file

In [15]:
books.duplicated().sum()

0

duplicate in file rating and users

In [16]:
users.duplicated().sum()

0

duplicate file in ratings

In [17]:
ratings.duplicated().sum()

0

popularity based recommmendation system. To show highest rated book..May be some book got over less ratings but high. So, that may effect the popularity. We will take into account only those books that got more than 250 ratings
Display the top 50 rated books

In [18]:

ratings

User-ID         ISBN  Book-Rating
0         276725   034545104X            0
1         276726   0155061224            5
2         276727   0446520802            0
3         276729   052165615X            3
4         276729   0521795028            6
...          ...          ...          ...
1149775   276704   1563526298            9
1149776   276706   0679447156            0
1149777   276709   0515107662           10
1149778   276721   0590442449           10
1149779   276723  05162443314            8

[1149780 rows x 3 columns]

In [19]:
books

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

Both ratings and books have merged based on ISBN column

In [20]:
ratings.merge(books,on='ISBN').shape

(1031136, 10)

In [21]:
ratings_with_name=ratings.merge(books,on='ISBN')

ratings with name holds the books data along with rating and dataframe

In [22]:
ratings_with_name

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1           2313  034545104X            5   
2           6543  034545104X            0   
3           8680  034545104X            5   
4          10314  034545104X            9   
...          ...         ...          ...   
1031131   276688  0517145553            0   
1031132   276688  1575660792            7   
1031133   276690  0590907301            0   
1031134   276704  0679752714            0   
1031135   276704  0806917695            5   

                                                Book-Title      Book-Author  \
0                                     Flesh Tones: A Novel       M. J. Rose   
1                                     Flesh Tones: A Novel       M. J. Rose   
2                                     Flesh Tones: A Novel       M. J. Rose   
3                                     Flesh Tones: A Novel       M. J. Rose   
4                                     Flesh Tones: A Novel       M. J. Rose   
...                                                    ...              ...   
1031131                                    Mostly Harmless    Douglas Adams   
1031132                                        Gray Matter  Shirley Kennett   
1031133  Triplet Trouble and the Class Trip (Triplet Tr...     Debbie Dadey   
1031134  A Desert of Pure Feeling (Vintage Contemporaries)   Judith Freeman   
1031135  Perplexing Lateral Thinking Puzzles: Scholasti...      Paul Sloane   

        Year-Of-Publication                          Publisher  \
0                      2002                   Ballantine Books   
1                      2002                   Ballantine Books   
2                      2002                   Ballantine Books   
3                      2002                   Ballantine Books   
4                      2002                   Ballantine Books   
...                     ...                                ...   
1031131                1995             Random House Value Pub   
1031132                1996  Kensington Publishing Corporation   
1031133                1997                              Apple   
1031134                1997                  Vintage Books USA   
1031135                1997                Sterling Publishing   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-L  
0        http://images.amazon.com/images/P/034545104X.0...  
1        http://images.amazon.com/images/P/034545104X.0...  
2        http://images.amazon.com/images/P/034545104X

Now calculate the average ratings of each books and consider only that book whose average rating is more than 250

In [23]:
num_rating_df=ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns={'Book-Rating':'num_rating'},inplace=True)
num_rating_df

Book-Title  num_rating
0        A Light in the Storm: The Civil War Diary of ...           4
1                                   Always Have Popsicles           1
2                    Apple Magic (The Collector's series)           1
3        Ask Lily (Young Women of Faith: Lily Series, ...           1
4        Beyond IBM: Leadership Marketing and Finance ...           1
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.           2
241067                     Ã?Â?rger mit Produkt X. Roman.           4
241068                                Ã?Â?sterlich leben.           1
241069                              Ã?Â?stlich der Berge.           3
241070                                  Ã?Â?thique en toc           2

[241071 rows x 2 columns]

The above will give us the books name with ratings on that books

In [24]:
avg_rating_df=ratings_with_name.groupby('Book-Title').mean()['Book-Rating'].reset_index()
avg_rating_df.rename(columns={'Book-Rating':'avg_rating'},inplace=True)
avg_rating_df

<ipython-input-24-454d828bd480>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  avg_rating_df=ratings_with_name.groupby('Book-Title').mean()['Book-Rating'].reset_index()


Book-Title  avg_rating
0        A Light in the Storm: The Civil War Diary of ...    2.250000
1                                   Always Have Popsicles    0.000000
2                    Apple Magic (The Collector's series)    0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...    8.000000
4        Beyond IBM: Leadership Marketing and Finance ...    0.000000
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.    0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.    5.250000
241068                                Ã?Â?sterlich leben.    7.000000
241069                              Ã?Â?stlich der Berge.    2.666667
241070                                  Ã?Â?thique en toc    4.000000

[241071 rows x 2 columns]

above prints the average ratings of the books

In [25]:
popularity_df=num_rating_df.merge(avg_rating_df,on='Book-Title')
popularity_df

Book-Title  num_rating  \
0        A Light in the Storm: The Civil War Diary of ...           4   
1                                   Always Have Popsicles           1   
2                    Apple Magic (The Collector's series)           1   
3        Ask Lily (Young Women of Faith: Lily Series, ...           1   
4        Beyond IBM: Leadership Marketing and Finance ...           1   
...                                                   ...         ...   
241066                                      Ã?Â?lpiraten.           2   
241067                     Ã?Â?rger mit Produkt X. Roman.           4   
241068                                Ã?Â?sterlich leben.           1   
241069                              Ã?Â?stlich der Berge.           3   
241070                                  Ã?Â?thique en toc           2   

        avg_rating  
0         2.250000  
1         0.000000  
2         0.000000  
3         8.000000  
4         0.000000  
...            ...  
241066    0.000000  
241067    5.250000  
241068    7.000000  
241069    2.666667  
241070    4.000000  

[241071 rows x 3 columns]

now we have to selecct only those books whose rating is higher than 250

In [26]:
popularity_df[popularity_df['num_rating']>=250].sort_values('avg_rating',ascending=False)

Book-Title  num_rating  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)         428   
80422        Harry Potter and the Goblet of Fire (Book 4)         387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)         278   
80426   Harry Potter and the Order of the Phoenix (Boo...         347   
80414    Harry Potter and the Chamber of Secrets (Book 2)         556   
...                                                   ...         ...   
227701       Vinegar Hill (Oprah's Book Club (Paperback))         265   
233635                                           Whispers         286   
143377                                  Presumed Innocent         294   
94382                                        Isle of Dogs         288   
234951                                        Wild Animus        2502   

        avg_rating  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453  
...            ...  
227701    2.245283  
233635    2.199301  
143377    2.139456  
94382     2.000000  
234951    1.019584  

[186 rows x 3 columns]

The above will give you the books that will select the books whose rating is greater than 250 and are in sorted order

In [27]:
popularity_df[popularity_df['num_rating']>=250].sort_values('avg_rating',ascending=False).head(50)

Book-Title  num_rating  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)         428   
80422        Harry Potter and the Goblet of Fire (Book 4)         387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)         278   
80426   Harry Potter and the Order of the Phoenix (Boo...         347   
80414    Harry Potter and the Chamber of Secrets (Book 2)         556   
191612  The Hobbit : The Enchanting Prelude to The Lor...         281   
187377  The Fellowship of the Ring (The Lord of the Ri...         368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...         575   
211384     The Two Towers (The Lord of the Rings, Part 2)         260   
219741                              To Kill a Mockingbird         510   
183573                                  The Da Vinci Code         898   
187880                 The Five People You Meet in Heaven         430   
180556                             The Catcher in the Rye         449   
196326                          The Lovely Bones: A Novel        1295   
764                                                  1984         284   
144165                           Prodigal Summer: A Novel         253   
128670                                         Neverwhere         265   
206502                            The Secret Life of Bees         774   
168719  Stupid White Men ...and Other Sorry Excuses fo...         283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...         493   
204387                The Red Tent (Bestselling Backlist)         723   
191589               The Hitchhiker's Guide to the Galaxy         268   
129379   Nickel and Dimed: On (Not) Getting By in America         335   
93381                                       Into the Wild         252   
63867                                      Fahrenheit 451         409   
74750                           Girl with a Pearl Earring         526   
136145                                          Outlander         283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...         585   
156102                     Seabiscuit: An American Legend         275   
107962                                         Life of Pi         664   
176845                                     The Bean Trees         389   
2281    A Child Called \It\": One Child's Courage to S...         265   
8434                                       ANGELA'S ASHES         279   
76343                                         Good in Bed         490   
64931   Fast Food Nation: The Dark Side of the All-Ame...         321   
12700                                       American Gods         302   
161645                                 Skipping Christmas         322   
105777  Left Behind: A Novel of the Earth's Last Days ...         318   
189551    The Golden Compass (His Dark Materials, Book 1)         336   
181679                                   The Color Purple         314   
160336                               Silence of the Lambs         256   
8752                                          About a Boy         262   
158138                  Seven Up (A Stephanie Plum Novel)         278   
175097  The Alchemist: A Fable About Following Your Dream         266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...         269   
170101                       Suzanne's Diary for Nicholas         457   
111073                                  Lord of the Flies         259   
5664                              A Prayer for Owen Meany         413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)         301   
233851       White Oleander : A Novel (Oprah's Book Club)         356   

        avg_rating  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453  
191612    5.007117  
187377    4.948370  
80445     4.895652  
211384    4.880769  
219741    4.700000  
183573    4.642539  
187880    4.551163  
180556    4.545657  
196326    4.468726  
764     

The above will give us the top-50 books

In [28]:
popular_books=popularity_df[popularity_df['num_rating']>=250].sort_values('avg_rating',ascending=False)
popular_books

Book-Title  num_rating  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)         428   
80422        Harry Potter and the Goblet of Fire (Book 4)         387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)         278   
80426   Harry Potter and the Order of the Phoenix (Boo...         347   
80414    Harry Potter and the Chamber of Secrets (Book 2)         556   
...                                                   ...         ...   
227701       Vinegar Hill (Oprah's Book Club (Paperback))         265   
233635                                           Whispers         286   
143377                                  Presumed Innocent         294   
94382                                        Isle of Dogs         288   
234951                                        Wild Animus        2502   

        avg_rating  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453  
...            ...  
227701    2.245283  
233635    2.199301  
143377    2.139456  
94382     2.000000  
234951    1.019584  

[186 rows x 3 columns]

In [29]:
popular_books.merge(books,on='Book-Title')

Book-Title  num_rating  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)         428   
1    Harry Potter and the Prisoner of Azkaban (Book 3)         428   
2    Harry Potter and the Prisoner of Azkaban (Book 3)         428   
3         Harry Potter and the Goblet of Fire (Book 4)         387   
4         Harry Potter and the Goblet of Fire (Book 4)         387   
..                                                 ...         ...   
735                                       Isle of Dogs         288   
736                                       Isle of Dogs         288   
737                                       Isle of Dogs         288   
738                                       Isle of Dogs         288   
739                                        Wild Animus        2502   

     avg_rating        ISBN                Book-Author Year-Of-Publication  \
0      5.852804  0439136350              J. K. Rowling                1999   
1      5.852804  0439136369              J. K. Rowling                2001   
2      5.852804  0786222743              J. K. Rowling                2000   
3      5.824289  0439139597              J. K. Rowling                2000   
4      5.824289  0439139600              J. K. Rowling                2002   
..          ...         ...                        ...                 ...   
735    2.000000  0399147659          Patricia Cornwell                2001   
736    2.000000  0425186768          Patricia Cornwell                2002   
737    2.000000  0739420704          Patricia Cornwell                2001   
738    2.000000  B0000C7BNG  Patricia Daniels Cornwell                2001   
739    1.019584  0971880107               Rich Shapero                2004   

                 Publisher                                        Image-URL-S  \
0               Scholastic  http://images.amazon.com/images/P/0439136350.0...   
1               Scholastic  http://images.amazon.com/images/P/0439136369.0...   
2          Thorndike Press  http://images.amazon.com/images/P/0786222743.0...   
3               Scholastic  http://images.amazon.com/images/P/0439139597.0...   
4    Scholastic Paperbacks  http://images.amazon.com/images/P/0439139600.0...   
..                     ...                                                ...   
735   Putnam Berkley Audio  http://images.amazon.com/images/P/0399147659.0...   
736      Berkley Pub Group  http://images.amazon.com/images/P/0425186768.0...   
737    G. P. Putnam's Sons  http://images.amazon.com/images/P/0739420704.0...   
738       Putnam Pub Group  http://images.amazon.com/images/P/B0000C7BNG.0...   
739                Too Far  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
1    http://images.amazon.com/images/P/0439136369.0...   
2    http://images.amazon.com/images/P/0786222743.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
4    http://images.amazon.com/images/P/0439139600.0...   
..                                                 ...   
735  http://images.amazon.com/images/P/0399147659.0...   
736  http://images.amazon.com/images/P/0425186768.0...   
737  http://images.amazon.com/images/P/0739420704.0...   
738  http://images.amazon.com/images/P/B0000C7BNG.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
1    http://images.amazon.com/images/P/0439136369.0...  
2    http://images.amazon.com/images/P/0786222743.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
4    http://images.amazon.com/images/P/0439139600.0...  
..                                                 ...  
735  http://images.amazon.com/images/P/0399147659.0...  
736  http://images.amazon.com/images/P/0425186768.0...  
737  http://images.amazon.com/images/P/0739420704.0...  
738  http://images.amazon.

the above will give us and we want the top-50 books author name, image and publisher is there..for that we need to merge the popular books with books. The rows are more that means there are duplicates e.g. harry potter book are coming multiple times with different ISBN

In [30]:
popular_books.merge(books,on='Book-Title').drop_duplicates('Book-Title')

Book-Title  num_rating  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)         428   
3         Harry Potter and the Goblet of Fire (Book 4)         387   
5       Harry Potter and the Sorcerer's Stone (Book 1)         278   
9    Harry Potter and the Order of the Phoenix (Boo...         347   
13    Harry Potter and the Chamber of Secrets (Book 2)         556   
..                                                 ...         ...   
716       Vinegar Hill (Oprah's Book Club (Paperback))         265   
717                                           Whispers         286   
727                                  Presumed Innocent         294   
733                                       Isle of Dogs         288   
739                                        Wild Animus        2502   

     avg_rating        ISBN        Book-Author Year-Of-Publication  \
0      5.852804  0439136350      J. K. Rowling                1999   
3      5.824289  0439139597      J. K. Rowling                2000   
5      5.737410  0590353403      J. K. Rowling                1998   
9      5.501441  043935806X      J. K. Rowling                2003   
13     5.183453  0439064872      J. K. Rowling                2000   
..          ...         ...                ...                 ...   
716    2.245283  0380730138   A. Manette Ansay                1998   
717    2.199301  0440216745        BELVA PLAIN                1994   
727    2.139456  0446359866        Scott Turow                1989   
733    2.000000  0425182908  Patricia Cornwell                2002   
739    1.019584  0971880107       Rich Shapero                2004   

                    Publisher  \
0                  Scholastic   
3                  Scholastic   
5                  Scholastic   
9                  Scholastic   
13                 Scholastic   
..                        ...   
716                 Perennial   
717                      Dell   
727              Warner Books   
733  Berkley Publishing Group   
739                   Too Far   

                                           Image-URL-S  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
5    http://images.amazon.com/images/P/0590353403.0...  
9    http://images.amazon.com/images/P/043935806X.0...  
13   http://images.amazon.com/images/P/0439064872.0...  
..                                                 ...  
716  http://images.amazon.com/images/P/0380730138.0...  
717  http://images.amazon.com/images/P/0440216745.0...  
727  http://images.amazon.com/images/P/0446359866.0...  
7

The above will give us all the information of the book..we can also select the info that we want as below

In [31]:
popular_books.merge(books,on='Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Publisher','num_rating','avg_rating']]

Book-Title        Book-Author  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)      J. K. Rowling   
3         Harry Potter and the Goblet of Fire (Book 4)      J. K. Rowling   
5       Harry Potter and the Sorcerer's Stone (Book 1)      J. K. Rowling   
9    Harry Potter and the Order of the Phoenix (Boo...      J. K. Rowling   
13    Harry Potter and the Chamber of Secrets (Book 2)      J. K. Rowling   
..                                                 ...                ...   
716       Vinegar Hill (Oprah's Book Club (Paperback))   A. Manette Ansay   
717                                           Whispers        BELVA PLAIN   
727                                  Presumed Innocent        Scott Turow   
733                                       Isle of Dogs  Patricia Cornwell   
739                                        Wild Animus       Rich Shapero   

                    Publisher  num_rating  avg_rating  
0                  Scholastic         428    5.852804  
3                  Scholastic         387    5.824289  
5                  Scholastic         278    5.737410  
9                  Scholastic         347    5.501441  
13                 Scholastic         556    5.183453  
..                        ...         ...         ...  
716                 Perennial         265    2.245283  
717                      Dell         286    2.199301  
727              Warner Books         294    2.139456  
733  Berkley Publishing Group         288    2.000000  
739                   Too Far        2502    1.019584  

[186 rows x 5 columns]

In [32]:
popular_books

Book-Title  num_rating  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)         428   
80422        Harry Potter and the Goblet of Fire (Book 4)         387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)         278   
80426   Harry Potter and the Order of the Phoenix (Boo...         347   
80414    Harry Potter and the Chamber of Secrets (Book 2)         556   
...                                                   ...         ...   
227701       Vinegar Hill (Oprah's Book Club (Paperback))         265   
233635                                           Whispers         286   
143377                                  Presumed Innocent         294   
94382                                        Isle of Dogs         288   
234951                                        Wild Animus        2502   

        avg_rating  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453  
...            ...  
227701    2.245283  
233635    2.199301  
143377    2.139456  
94382     2.000000  
234951    1.019584  

[186 rows x 3 columns]

**This is how we will build the popularity based recommender systems..Here the concept is we are considering the average rating where minimum rating should be greater than 250**

**For collaborative filtering we need a file which contains the books and the rating provided by each user e.g. book1=user1|user2|user3|...|usern**
**we will consider only those user whose has given more than 200 ratings that means teh user who gives ratings more often..and the books only those who got more than 50 rating**

** Collaborative Filteirng Recommender System **

In [33]:
ratings_with_name.groupby('User-ID').count()['Book-Rating']>200

User-ID
2         False
8         False
9         False
10        False
12        False
          ...  
278846    False
278849    False
278851    False
278852    False
278854    False
Name: Book-Rating, Length: 92106, dtype: bool

In [34]:
x=ratings_with_name.groupby('User-ID').count()['Book-Rating']>200
Active_User=x[x].index

The above will give user id who rated more than 200 books

In [35]:
filtered_rating=ratings_with_name[ratings_with_name['User-ID'].isin(Active_User)]
filtered_rating

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.a

From all the users we are selecting the info of the active users

In [36]:
filtered_rating.groupby('Book-Title').count()['Book-Rating']

Book-Title
 A Light in the Storm: The Civil War Diary of Amelia Martin, Fenwick Island, Delaware, 1861 (Dear America)    2
 Always Have Popsicles                                                                                        1
 Apple Magic (The Collector's series)                                                                         1
 Beyond IBM: Leadership Marketing and Finance for the 1990s                                                   1
 Clifford Visita El Hospital (Clifford El Gran Perro Colorado)                                                1
                                                                                                             ..
Ã?Â?ber das Fernsehen.                                                                                        2
Ã?Â?ber die Pflicht zum Ungehorsam gegen den Staat.                                                           3
Ã?Â?lpiraten.                                                                                

So we will get by above is book and no of rating on the book

In [37]:
y=filtered_rating.groupby('Book-Title').count()['Book-Rating']>=50
famous_books=y[y].index

users who has rated minimum 50 books has been selected in famous books..

In [38]:
famous_books

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

now we are checking that these famous books are in filtered book and getting the details

In [39]:
final_rating=filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]

This is our final rating..Corresponding to every book which user has rated

In [40]:
final_rating

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [41]:
final_rating.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

This is the format that is require where user who rated above 200 and book above 50 rating

In [42]:
pt=final_rating.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')

In [43]:
pt.fillna(0,inplace=True)
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

Book1 is represented by 810 dimension say book1 is a vector of 810 dimesnion space..each book is a vector in 810 dimensional space..here we have to find similar vector..we will calculate the similarity score for each vector with every other vectors that will give us the similarity between the two books..

In [44]:
from sklearn.metrics.pairwise import cosine_similarity

In [45]:
similarity_score=cosine_similarity(pt)

we are getting 2D with shape 706 by 706 that means we are getting 706 books similarity with other 706 books

similarity_score

In [46]:
similarity_score[0]

array([1.        , 0.10255025, 0.01220856, 0.        , 0.05367224,
       0.02774901, 0.08216491, 0.13732869, 0.03261686, 0.03667591,
       0.02322418, 0.06766487, 0.02083978, 0.09673735, 0.13388865,
       0.08303112, 0.11153543, 0.05100411, 0.02517784, 0.11706383,
       0.        , 0.14333793, 0.07847534, 0.06150451, 0.08723968,
       0.        , 0.07009814, 0.13658681, 0.07600328, 0.12167134,
       0.00768046, 0.01473221, 0.        , 0.07965814, 0.04522617,
       0.01556271, 0.09495938, 0.0182307 , 0.02610465, 0.07984012,
       0.11679969, 0.0569124 , 0.08354155, 0.08471898, 0.08785938,
       0.05491435, 0.0548505 , 0.27026514, 0.09779123, 0.06016046,
       0.08958835, 0.06748675, 0.        , 0.04468098, 0.01920872,
       0.        , 0.05629067, 0.00557964, 0.07877059, 0.05219479,
       0.18908177, 0.        , 0.01240656, 0.02984572, 0.04279502,
       0.12680125, 0.16566735, 0.        , 0.13357242, 0.06615478,
       0.        , 0.        , 0.        , 0.10968075, 0.02806

This is the first row that tells us that first book 1984 similarity with other books

In [47]:
similarity_score.shape

(706, 706)

now we will make a function recommend that will take one book and it will suggest 5 books similar to the book

In [48]:
def recom(book_name):
  # here we have to fetch the index of the book that has been passed as
  index=np.where(pt.index==book_name)[0][0]
  #once index is fetched then go that index in similarity score
  similar_items=sorted(list(enumerate(similarity_score[index])),key=lambda x:x[1],reverse=True)[1:6]
  for i in similar_items:
    print(pt.index[i[0]])

**The top-5 movies similar to given movies**

In [49]:
recom('1984')

Animal Farm
The Handmaid's Tale
Brave New World
The Vampire Lestat (Vampire Chronicles, Book II)
The Hours : A Novel
